dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,607 @@
 package api
 
 import (
-	"k8s.io/apimachinery/pkg/api/resource"
+	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
 )
 
 // DeviceClass is a vendor or admin-provided resource that contains
-// contraint and configuration information. Essentially, it is a re-usable
-// collection of predefined data that device claims may use.
+// device configuration and requirements. It can be referenced in
+// the device requests of a claim to apply these presets.
 // Cluster scoped.
 type DeviceClass struct {
-	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata,omitempty"`
-
-	Spec DeviceClassSpec `json:"spec,omitempty"`
-}
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
 
-// DeviceClassSpec provides either a selector to find other
-// classes, or the specific constraints defining this class.
-type DeviceClassSpec struct {
+	// Only nodes matching the selector will be considered by the scheduler
+	// when trying to find a Node that fits a Pod when that Pod uses
+	// a ResourceClaim that has not been allocated yet.
+	//
+	// Setting this field is optional. If unset, all nodes are candidates.
+	// It is ignored when the claim does not use a control plane controller
+	// for allocation.
+	//
+	// This is an alpha field and requires enabling the DRAControlPlaneController
+	// feature gate.
+	//
+	// +optional
+	SuitableNodes *v1.NodeSelector `json:"suitableNodes,omitempty" protobuf:"bytes,4,opt,name=suitableNodes"`
 
-	// One of selector and classCriteria must be populated.
+	// Config defines configuration parameters that apply to each request in a claim using this class.
+	// They are ignored while allocating the claim.
+	//
+	// +optional
+	// +listType=atomic
+	Config []ConfigurationParameters `json:"config,omitempty"`
 
-	// Selector specifies a label selector used to identify classes which
-	// may be considered equivalent to this class. In other words, this
-	// allows the adminstrator to define groups of classes which may be
-	// referred to as a single class in the claim.
+	// Requirements must be satisfied by devices. Applies to all devices of
+	// a claim when the claim references the class and only to the devices
+	// in a request when referenced there.
 	//
 	// +optional
-	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+	// +listType=atomic
+	Requirements []Requirement `json:"requirements,omitempty" protobuf:"bytes,4,opt,name=requirements"`
```

Comment:
"Constraints"?

Reply:
@klueska and I found it useful to have different terms for "constraints" (among sets of devices) and "requirements" (for a single device).

As it stands now, we have `Requirements []Requirement` and `Constraints []Constraint`. If we were to rename, we would end up with `Requirements []RequestRequirements` (here and in `ResourceRequestDetail`) and `Requirements []ClaimRequirements` (in `ResourceClaimSpec`).

I don't care too strongly.

Reply:
I think that's a good idea, to differentiate between constraints on the individual devices versus constraint on a chosen set of ideas.

Reply:
As I understand the proposal, requirements are what we used to call constraints, and constraints are the conditions that a selected set of devices must satisfy, internally. This comes down to that we are allocating not just a *single device* but instead a *set of devices*. Requirements apply to devices; constraints apply to the set of devices. That is:

- Requirements general filter criteria for the possible devices that might satisfy the claim. Only devices that match all requirements will even be considered in the solution. Examples:
  - Devices of vendor "foo"
  - Devices of model "bar"
  - Devices with more than 80Gi of memory
- Constraints are on top of the chosen set of devices. Examples:
  - I need 1-4 devices
  - Devices must in aggregate have 120Gi of memory.
  - I need exactly 3 devices
  - Devices must all have these matching attributes
 
Another way to think of it is that a requirement is equivalent to a function `func my_requirement(Device) bool` where the passed in device is any device, whereas a constraint is a equivalent to a function `func my_constraint([]Device) bool` where the passed in []Device is a possible solution set of devices already passing requirements (and thus requirement-style logic would be frowned upon in there).

In 1.31, I am suggesting that we actually do NOT allow allocating a set of devices in a single request, just to cut scope (since asking for N devices is equivalent to repeating a request N times). Thus, we will have constraints across the devices in a claim, but not in a request, and the only one we will implement is "match attributes". If we instead choose to implement multiple devices per request, we would also need an "exact count" constraint to tell us how many in the request, and a "match attributes" for the request scope.

Reply:
The "count" and "per-request constraint" are now a "FUTURE EXTENSION". I kept "claim.constraints" with an extended comment that mentions "set of devices".

Reply:
Some outsider thoughts on words for starting individual device requirements (A) vs further refinements/aggregate constraints (B).

It seems like the desired direction is away from qualifying adjectives, but if we don't mind them, something in the vein of:
A: BaseRequirements / DeviceRequirements
B: GroupRequirements / PoolRequirements

Pairs leaning more on different words:
A: DeviceMinimum / Minimum, Start / Initial, Base, Strict
B: Perimeter, Limit

Reply:
The intent was indeed to group several different criteria together with an implied AND. Future criteria would be consuming some resource.

Another thought was that allowing more than one selector might be useful. Semantically that's the same as a larger CEL expression with `&&`, so we do not absolutely need this.

Reply:
As we don't have consensus on "constraints" vs. "requirements" I've not changed that yet.

Reply:
Let's just put it on the to do list and not hold the PR.

Reply:
I created https://github.com/kubernetes-sigs/wg-device-management/issues/16

Reply:
I am skeptical that there will be another form of criteria, and doubly skeptical that we would want to allow BOTH at the same time.  So if this became `constraints string` or even `[]string` I'd be super happy.

Reply:
This is one of those areas where I think a one-of is useful. It might be unlikely that we add some other criteria, but if we ever need to, then currently it is safe to do so.

If we make this a `constraints string` or array, then we have to allow it to be empty because the class might be sufficient. That means a current scheduler will happily proceed with allocating a claim even though it is ignoring some future field.

To me, this is a risk assessment: `<probability> * <negative impact>`. The probability may be low, but the impact would be big.

Reply:
An argument against `string` and in favor of a struct is that some CEL-ish APIs offer variable definitions alongside the expression.

Reply:
_IF_ we believe that we may one day want a totally different form of criteria, then we probably want to ADD a level of structure.

1) Do you think there's a need for more than one CEL expression specified in this type?
2) Do you think there will be other (non-CEL) constraints?
3) Do you think we will want to support CEL and non-CEL at the same time?

If you buy the argument above for variable definitions, then a CEL expression is a struct, not a string.

The simplest would be `constraints CELExpr`.

If we need more than 1 CEL expr: `constraints []CELExpr`

If we need other non-CEL:
```
  // one of these
  celConstraints *CELExpr
  otherConstraints *OtherArgs
```

If we need multiple constraints: `constraints []ConstraintOption`, which holds a CELExpr.  It makes me sad if we go there, but I understand it now.

Reply:
I don't see resource allocations as constraints, so I wouldn't include those in `Requirements` but instead they would be part of some sort of `Requests` field that looks like ordinary resource requests (ie, is a `map[string]resource.Quantity`). I think doing allocatable requests in a different way from Pod would be a confusing UX.

Also, this type is used in both class and claim. I don't think resource requests belong in class. Maybe I just have too rigid of a mental model :)

I guess maybe we would want to have some non-CEL thing like simplistic equality constraints if people don't want to write CEL. That would be an alternative. But that just seems like two ways to do the same thing.

However I will say that if we are creating a struct anyway to capture CEL variable names or other sort of future CEL magic, I don't see the point in avoiding the oneOf. That is, if we want `CELExpr` not just `string`, we may as well have `ConstraintOption`, IMO.

Reply:
We should continue this discussion on https://github.com/kubernetes-sigs/wg-device-management/issues/16 not here, please. This PR is merged and ...enormous.